### PR TITLE
fixup sloppy cache implementation

### PR DESCRIPTION
**Pull request by maxtaco**
- should allow an uncached implemention
- should initialize with other caches in `GlobalContext#ConfigureCaches`
- provide a generic UPAKLoader interface
  - by default no caching
  - but we turn it on after ConfigureCaches
  - also make it thread safe w/r/t resetting caches
  - don't use cache in nocache cases
  - rename cached_user_loader to upak_loader
- full self cacher is now also thread-safe, and works uncached.

### CODE REVIEW

**Comment by maxtaco** [comment]
Also add to the above rooting out the `UserDeviceCache`.  Since we added the `CachedUserLoader`, it's not accomplishing much.

**Comment by maxtaco** [comment]
BTW, it's OK with @mmaxim to nuke the `UserDeviceCache` (as per in-person convo).

File: go/libkb/globals.go [hunk 1]
```diff
@@ -176,9 +175,8 @@ func (g *GlobalContext) Init() *GlobalContext {
 	g.createLoginState()
 	g.Resolver = NewResolver(g)
 	g.RateLimits = NewRateLimits(g)
-	g.CachedUserLoader = NewCachedUserLoader(g, CachedUserTimeout)
-	g.UserDeviceCache = NewUserDeviceCache(g)
-	g.makeFullSelfCacher()
+	g.upakLoader = NewUncachedUPAKLoader(g)
+	g.fullSelfer = NewNullFullSelfCache(g)
```
Comment (patrickxb) [inline go/libkb/globals.go]: why is this a NullFullSelfCache?

Comment (maxtaco) [inline go/libkb/globals.go]: the idea is that it won't crash, but it won't cache if you were to run it after the global context object initializes.  Later, when we {{ConfigureCaches}}, we get a real implementation swapped in.  I agree the name is bad, should I rename it?  `NewUncachedFullSelfer` ?

Comment (patrickxb) [inline go/libkb/globals.go]: Sounds good,  would match the line above it.

I didn't see the initialization in ConfigureCaches in the gh diff.

**Comment by juzam** [comment]
Maybe I'm doing something wrong but this PR broke the build of the client for me.
I'm building it with golang 1.7.4 via an arm docker image built with this dockerfie: https://github.com/juzam/golang-arm-dockerfiles/blob/master/Docker

here's the relevant output:

```
Step 8 : RUN go get github.com/keybase/client/go/keybase && go build -tags production github.com/keybase/client/go/keybase
 ---> Running in fbf53ebc001c
# github.com/keybase/client/go/engine
src/github.com/keybase/client/go/engine/identify2_with_uid.go:131: g.CachedUserLoader undefined (type *libkb.GlobalContext has no field or method CachedUserLoader)
```

**Comment by maxtaco** [comment]
No, I blew it, I should have rebased and reran tests before this merge.  Should be fixed now in master, sorry about that...

**Comment by juzam** [comment]
No problem. I must have pulled the repository just before the fix. Everything is fine now, thank you!

**Comment by maxtaco** [comment]
My attempt to make a hotfix might not have worked, so I'm going to back out this PR and try again once the dust settles.

**Pull request merged**